cli/src/commands/run.rs

```diff
@@ -14,15 +14,375 @@
+    if !base_path.exists() {
+        fs::create_dir_all(&base_path)?;
+    }
+    tracing::debug!(path = ?base_path, "creating working copies in path: ");
+    for commit in commits {
+        let name = to_wc_name(commit.tree_id());
+        let commit_path = base_path.join(name.as_str());
+        fs::create_dir(&commit_path)?;
+        tracing::debug!(
+            dir = ?commit_path,
+            commit = commit.id().hex(),
+            "creating directory for the commit"
+        );
+
+        let (output_dir, working_copy_dir, state_dir) = create_working_copy_paths(&commit_path)?;
+        let tree_state = {
+            tracing::debug!(
+                commit = commit.id().hex(),
+                "trying to create a treestate for commit"
+            );
+            let mut tree_state = TreeState::init(
+                commit.store().clone(),
+                working_copy_dir.clone(),
+                state_dir.clone(),
+            )?;
+            tree_state
+                .check_out(
+                    &commit.tree()?,
+                    &CheckoutOptions {
+                        conflict_marker_style: ConflictMarkerStyle::Diff,
+                    },
+                )
+                .map_err(|_| RunError::FailedCheckout)?;
+            Mutex::new(tree_state)
+        };
+        let stored_commit = OnDiskCommit::new(
+            Some(name),
+            commit,
+            output_dir,
+            working_copy_dir,
+            state_dir,
+            tree_state,
+        );
+        results.push(Arc::new(stored_commit));
+    }
+    Ok(results)
+}
+
+/// Get the shell to execute in.
+// TODO: use something like `[run].shell`
+fn get_shell_executable() -> String {
+    if cfg!(target_os = "windows") {
+        "cmd".into()
+    } else {
+        "/bin/sh".into()
+    }
+}
+
+/// The result of a single command invocation.
+struct RunJob {
+    /// The old `CommitId` of the commit.
+    old_id: CommitId,
+    /// The new tree generated from the commit.
+    new_tree: Tree,
+    /// Was the tree even modified.
+    dirty: bool,
+}
+
+// TODO: make this more revset/commit stream friendly.
+async fn run_inner<'a>(
+    tx: &WorkspaceCommandTransaction<'a>,
+    sender: Sender<RunJob>,
+    handle: &tokio::runtime::Handle,
+    shell_command: Arc<String>,
+    commits: Arc<Vec<Arc<OnDiskCommit>>>,
+) -> Result<(), RunError> {
+    let mut command_futures = JoinSet::new();
+    for commit in commits.iter() {
+        command_futures.spawn_on(
+            rewrite_commit(
+                // TODO: handle error here
+                tx.base_workspace_helper().base_ignores().unwrap().clone(),
+                ConflictMarkerStyle::Snapshot,
+                commit.clone(),
+                shell_command.clone(),
+            ),
+            handle,
+        );
+    }
+
+    while let Some(res) = command_futures.join_next().await {
+        let done = res.unwrap().expect("should not fail joining a job");
```

Comment:
I ran into a panic here when testing locally with `jj run 'sed/some/string/g <file>'`, as the last file was unaffected which returned a exitcode of 255 (2).